### ISSUE TITLE
从优化算法角度看，像layers.elementwise_max, layers.clip这些不可导的op是如何被计算导数，并参与梯度优化过程的？

### ISSUE THREAD

**Issue by ydchen9** [PaddlePaddle/Paddle#22723]
比如下面这些代码，来自如下链接代码中的182行：https://github.com/PaddlePaddle/PARL/blob/develop/parl/algorithms/fluid/ppo.py

            pg_ratio = layers.exp(logprob - old_logprob)
            clipped_pg_ratio = layers.clip(pg_ratio, 1 - self.epsilon,
                                           1 + self.epsilon)
            surrogate_loss = layers.elementwise_min(
                advantages * pg_ratio, advantages * clipped_pg_ratio)
            loss = 0 - layers.reduce_mean(surrogate_loss)

**Comment by jiweibo** [comment]
可以参考这些op的反向实现。
https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/elementwise/elementwise_max_op.h
https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/clip_op.h

**Comment by ydchen9** [comment]
谢谢！
&nbsp;

祝您阖家幸福，工作顺利！
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 
 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;


&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Wilber"<notifications@github.com&gt;; 
发送时间: 2020年2月22日(星期六) 上午10:26
收件人: "PaddlePaddle/Paddle"<Paddle@noreply.github.com&gt;; 
抄送: "陈远东"<chenyd49@qq.com&gt;; "Author"<author@noreply.github.com&gt;; 
主题: Re: [PaddlePaddle/Paddle] 从优化算法角度看，像layers.elementwise_max, layers.clip这些不可导的op是如何被计算导数，并参与梯度优化过程的？ (#22723)




可以参考这些op的反向实现。
 https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/elementwise/elementwise_max_op.h
 https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/clip_op.h
